modules/dcache/src/main/java/org/dcache/services/ssh2/Ssh2Admin.java

```diff
@@ -318,6 +331,71 @@ public boolean authenticate(String userName, PublicKey key,
         }
     }
 
+    private static class RemoteHostValidator {
+
+        private enum Outcome {
+            ALLOW, DENY, DEFER
+        }
+
+        private boolean isValidHost(String line, ServerSession session) {
+            for (String linePart : Splitter.on(" ").trimResults().omitEmptyStrings().split(line)) {
+                if (linePart.startsWith("from=") && linePart.endsWith("\"")) {
+                    Set<Outcome> outcomes = EnumSet.noneOf(Outcome.class);
+                    for (String pattern : Splitter.on(",").trimResults().omitEmptyStrings().split(strip(linePart))) {
+                        outcomes.add(patternMatchesHost(pattern, session));
+                    }
+                    return outcomes.contains(Outcome.ALLOW) && !outcomes.contains(Outcome.DENY);
+                }
+            }
+            return true;
+        }
+
+        private Outcome patternMatchesHost(String pattern, ServerSession session) {
+            boolean patternMatches = false;
+            boolean patternIsNegated = pattern.startsWith("!");
+            if (patternIsNegated) {
+                pattern = pattern.substring(1);
+            }
+            SocketAddress remote = session.getClientAddress();
+            if (remote instanceof InetSocketAddress) {
+                String remoteAddress = ((InetSocketAddress) remote).getAddress().getHostAddress();
+                try {
+                    patternMatches = addressInCidrRange(pattern, remoteAddress);
+                } catch (IllegalArgumentException e) {
+                    String remoteName = ((InetSocketAddress) remote).getHostName();
+                    if (Glob.isGlob(pattern)) {
+                        Pattern p = Glob.parseGlobToPattern(pattern);
+                        Matcher addressMatcher = p.matcher(remoteAddress);
+                        Matcher nameMatcher = p.matcher(remoteName);
+                        patternMatches = (addressMatcher.matches() || nameMatcher.matches());
+                    } else {
+                        patternMatches = pattern.equals(remoteName);
+                    }
+                }
+            }
+            if (patternIsNegated && patternMatches) {
+                return Outcome.DENY;
+            } else if (!patternIsNegated && patternMatches) {
+                return Outcome.ALLOW;
+            } else {
+                return Outcome.DEFER;
+            }
+        }
+
+        private boolean addressInCidrRange(String cidrAddress, String remoteAddress) throws IllegalArgumentException {
+                try {
+                    Subnet subnet = Subnet.create(cidrAddress);
+                    return subnet.containsHost(remoteAddress);
+                } catch (UnknownHostException e) {
+                    return false;
+                }
+            }
+
+        private String strip(String linePart) {
+            return linePart.substring(6, linePart.length()-1);
```

Comment:
Not blocking the patch, but given this method is so small, why not include it the one place it gets called?
Something like:

    if (linePart.startsWith("from=") && linePart.endsWith("\"")) {
        String from = linePart.substring(6, linePart.length()-1);